Mathlib/Geometry/Manifold/Algebra/SmoothGerm.lean

```diff
@@ -0,0 +1,131 @@
+lemma mem_smoothGerm {x : M} (a : Germ (𝓝 x) N) :
+    a ∈ smoothGerm I I' N x ↔ ∃ f : SmoothMap I I' M N, Germ.ofFun f = a := by
+  rfl
+
+namespace smoothGerm
+instance (x : M) [ChartedSpace H' N] : Coe C^∞⟮I, M; I', N⟯ (smoothGerm I I' N x) :=
+  ⟨fun f ↦ ⟨(f : M → N), ⟨f, rfl⟩⟩⟩
+
+@[simp]
+theorem coe_coe (f : C^∞⟮I, M; I', N⟯) (x : M) :
+    ((f : smoothGerm I I' N x) : (𝓝 x).Germ N) = (f : (𝓝 x).Germ N) :=
+  rfl
+
+theorem coe_eq_coe (f g : C^∞⟮I, M; I', N⟯) {x : M} (h : ∀ᶠ y in 𝓝 x, f y = g y) :
+    (f : smoothGerm I I' N x) = (g : smoothGerm I I' N x) := by
+  ext
+  rwa [Germ.coe_eq]
+
+-- If `R` has the appropriate structure, `smoothGerm I I' R x` is a subgroup etc.
+-- All respective axioms are easy to prove by choosing explicit representatives.
+section subring
+
+-- xxx: would this be useful? Zero resp. One imply the space of smooth germs has one...
+variable (I' : ModelWithCorners 𝕜 E' H')
+  (R : Type*) [CommRing R] [TopologicalSpace R] [ChartedSpace H' R]
+
+/-- If `R` is a manifold with smooth multiplication,
+`smoothGerm I I' R x` is a sub-semigroup of `Germ (𝓝 x) R`. -/
+-- FIXME: is this definition useful, given it has the same assumptions as `submonoid`?
+def subsemigroup [SmoothMul I' R] (x : M) : Subsemigroup (Germ (𝓝 x) R) where
+  carrier := smoothGerm I I' R x
+  mul_mem' ha hb := by
+    choose f hf using ha
+    choose g hg using hb
+    exact ⟨f * g, by rw [← hf, ← hg, SmoothMap.coe_mul, Germ.coe_mul]⟩
+
+/-- If `R` is a manifold with smooth multiplication,
+`smoothGerm I I' R x` is a submonoid of `Germ (𝓝 x) R`. -/
+def submonoid [SmoothMul I' R] (x : M) : Submonoid (Germ (𝓝 x) R) where
+  toSubsemigroup := smoothGerm.subsemigroup I I' R x
+  one_mem' := ⟨1, by rw [SmoothMap.coe_one, Germ.coe_one]⟩
+
+/-- If `R` is a manifold with smooth addition,
+`smoothGerm I I' R x` is an additive sub-semigroup of `Germ (𝓝 x) R`. -/
+def addSubsemigroup [SmoothAdd I' R] (x : M) : AddSubsemigroup (Germ (𝓝 x) R) where
+  carrier := smoothGerm I I' R x
+  add_mem' ha hb := by
+    choose f hf using ha
+    choose g hg using hb
+    exact ⟨f + g, by rw [← hf, ← hg, SmoothMap.coe_add, Germ.coe_add]⟩
+
+/-- If `G` is an additive Lie group, `smoothGerm I I' G x` is
+an additive subgroup of `Germ (𝓝 x) G`. -/
+def addSubgroup [LieAddGroup I' R] (x : M) : AddSubgroup (Germ (𝓝 x) R) where
+  __ := smoothGerm.addSubsemigroup I I' R x
+  zero_mem' := ⟨0, by rw [SmoothMap.coe_zero, Germ.coe_zero]⟩
+  neg_mem' h := by
+    choose f hf using h
+    exact ⟨-f, by rw [← hf, SmoothMap.coe_neg, Germ.coe_neg]⟩
+
+/-- If `R` is a smooth ring, `smoothGerm I I' R x` is a subring of `Germ (𝓝 x) R`. -/
+def subring [SmoothRing I' R] (x : M) : Subring (Germ (𝓝 x) R) where
+  __ := smoothGerm.submonoid I I' R x
+  __ := smoothGerm.addSubgroup I I' R x
+
+@[simp, norm_cast]
+lemma coe_toSubring [SmoothRing I' R] (x : M) : smoothGerm.subring I I' R x = smoothGerm I I' R x :=
+  rfl
+
+lemma mem_toSubring [SmoothRing I' R] {x : M} (a : Germ (𝓝 x) R) :
```

Comment:
If you're going to having each intermediate structure, there should be corresponding lemmas for each.

Reply:
Fair point - added.